File: playwright-java/src/main/java/testtracks/runtime/BrowserSession.java
package testtracks.runtime;

import com.microsoft.playwright.Browser;
import com.microsoft.playwright.BrowserContext;
import com.microsoft.playwright.BrowserType;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.Playwright;
import com.microsoft.playwright.PlaywrightException;
import java.io.IOException;
import java.io.UncheckedIOException;
import java.net.URI;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import testtracks.configuration.BrowserName;
import testtracks.configuration.TestSettings;
import testtracks.diagnostics.UiEvidence;

public final class BrowserSession implements AutoCloseable {
  private final Playwright playwright;
  private final Browser browser;
  private final BrowserContext context;
  private final Page page;
  private final TestSettings settings;
  private final UiEvidence evidence;
  private boolean tracingStarted;

  private BrowserSession(
      Playwright playwright,
      Browser browser,
      BrowserContext context,
      Page page,
      TestSettings settings,
      UiEvidence evidence) {
    this.playwright = playwright;
    this.browser = browser;
    this.context = context;
    this.page = page;
    this.settings = settings;
    this.evidence = evidence;
  }

  public Page page() {
    return page;
  }

  public UiEvidence evidence() {
    return evidence;
  }

  public static BrowserSession create(TestSettings settings, URI baseUrl, String scenarioId) {
    var playwright = Playwright.create();
    var browserType =
        settings.browser() == BrowserName.FIREFOX ? playwright.firefox() : playwright.chromium();
    var launchOptions = new BrowserType.LaunchOptions().setHeadless(settings.headless());

    if (settings.browser() == BrowserName.EDGE) {
      launchOptions.setChannel("msedge");
    }

    try {
      var browser = browserType.launch(launchOptions);
      var context =
          browser.newContext(new Browser.NewContextOptions().setBaseURL(baseUrl.toString()));
      context.setDefaultTimeout(settings.defaultTimeoutMs());
      context.setDefaultNavigationTimeout(settings.defaultTimeoutMs());

      var page = context.newPage();
      page.setDefaultTimeout(settings.defaultTimeoutMs());
      page.setDefaultNavigationTimeout(settings.defaultTimeoutMs());

      var evidence = new UiEvidence();
      evidence.attachTo(page);
      var session = new BrowserSession(playwright, browser, context, page, settings, evidence);

      if (settings.traceOnFailure()) {
        context
            .tracing()
            .start(
                new com.microsoft.playwright.Tracing.StartOptions()
                    .setScreenshots(true)
                    .setSnapshots(true)
                    .setSources(true)
                    .setTitle("test-tracks-playwright-java " + scenarioId));
        session.tracingStarted = true;
      }

      return session;
    } catch (PlaywrightException ex) {
      playwright.close();
      if (settings.browser() == BrowserName.EDGE) {
        throw new IllegalStateException(
            "BROWSER=edge requires the Microsoft Edge stable channel. Install it with the Playwright browser install command for msedge.",
            ex);
      }
      throw ex;
    } catch (RuntimeException ex) {
      playwright.close();
      throw ex;
    }
  }

  public Path stopTracing(boolean saveTrace, String scenarioArtifactName) {
    if (!tracingStarted) {
      return null;
    }

    tracingStarted = false;
    if (!saveTrace) {
      context.tracing().stop();
      return null;
    }

    var tracePath = Path.of(settings.artifactsDirectory(), scenarioArtifactName, "trace.zip");
    try {
      Files.createDirectories(tracePath.getParent());
    } catch (IOException ex) {
      throw new UncheckedIOException("Could not create Playwright trace folder.", ex);
    }
    context.tracing().stop(new com.microsoft.playwright.Tracing.StopOptions().setPath(tracePath));
    return tracePath;
  }

  @Override
  public void close() {
    var failures = new ArrayList<RuntimeException>();

    // Keep closing later resources even when an earlier Playwright resource fails to dispose.
    closeCollecting("page", page::close, failures);
    closeCollecting("browser context", context::close, failures);
    closeCollecting("browser", browser::close, failures);
    closeCollecting("Playwright", playwright::close, failures);

    if (!failures.isEmpty()) {
      var failure =
          new IllegalStateException("One or more browser session resources failed to close.");
      failures.forEach(failure::addSuppressed);
      throw failure;
    }
  }

  private static void closeCollecting(
      String resourceName, Runnable action, List<RuntimeException> failures) {
    try {
      action.run();
    } catch (RuntimeException ex) {
      failures.add(new IllegalStateException("Failed to close " + resourceName + ".", ex));
    }
  }
}

File: playwright-csharp/src/Runtime/BrowserSession.cs
using Microsoft.Playwright;
using TestTracks.Playwright.CSharp.Configuration;
using TestTracks.Playwright.CSharp.Diagnostics;

namespace TestTracks.Playwright.CSharp.Runtime;

/// <summary>
/// Wraps a browser, browser context and page for one UI scenario.
/// </summary>
public sealed class BrowserSession : IAsyncDisposable
{
    private bool _tracingStarted;

    private BrowserSession(
        IPlaywright playwright,
        IBrowser browser,
        IBrowserContext context,
        IPage page,
        TestSettings settings,
        UiEvidence evidence)
    {
        Playwright = playwright;
        Browser = browser;
        Context = context;
        Page = page;
        Settings = settings;
        Evidence = evidence;
    }

    public IPlaywright Playwright { get; }

    public IBrowser Browser { get; }

    public IBrowserContext Context { get; }

    public IPage Page { get; }

    public TestSettings Settings { get; }

    public UiEvidence Evidence { get; }

    /// <summary>
    /// Starts a browser session against the supplied site base URL.
    /// </summary>
    /// <remarks>
    /// Each call creates a fresh browser context, so cookies, storage and page state are not shared between scenarios.
    /// </remarks>
    public static async Task<BrowserSession> CreateAsync(TestSettings settings, Uri baseUrl, string scenarioId)
    {
        Directory.CreateDirectory(settings.ArtifactsDirectory);

        var playwright = await Microsoft.Playwright.Playwright.CreateAsync();
        var browserType = settings.Browser == BrowserName.Firefox
            ? playwright.Firefox
            : playwright.Chromium;

        var launchOptions = new BrowserTypeLaunchOptions
        {
            Headless = settings.Headless
        };

        if (settings.Browser == BrowserName.Edge)
        {
            launchOptions.Channel = "msedge";
        }

        try
        {
            var browser = await browserType.LaunchAsync(launchOptions);
            var context = await browser.NewContextAsync(new BrowserNewContextOptions
            {
                BaseURL = baseUrl.ToString()
            });

            context.SetDefaultTimeout(settings.DefaultTimeoutMs);
            context.SetDefaultNavigationTimeout(settings.DefaultTimeoutMs);

            var page = await context.NewPageAsync();
            page.SetDefaultTimeout(settings.DefaultTimeoutMs);
            page.SetDefaultNavigationTimeout(settings.DefaultTimeoutMs);

            var evidence = new UiEvidence();
            evidence.AttachTo(page);

            var session = new BrowserSession(playwright, browser, context, page, settings, evidence);

            if (settings.TraceOnFailure)
            {
                await context.Tracing.StartAsync(new TracingStartOptions
                {
                    Screenshots = true,
                    Snapshots = true,
                    Sources = true,
                    Title = $"test-tracks-playwright-csharp {scenarioId}"
                });
                session._tracingStarted = true;
            }

            return session;
        }
        catch (PlaywrightException ex) when (settings.Browser == BrowserName.Edge)
        {
            playwright.Dispose();
            throw new InvalidOperationException(
                "browser=edge requires the Microsoft Edge stable channel. Install it with the Playwright browser install command for msedge.",
                ex);
        }
        catch
        {
            playwright.Dispose();
            throw;
        }
    }

    /// <summary>
    /// Stops Playwright tracing and optionally writes the trace zip to the run artifacts folder.
    /// </summary>
    public async Task<string?> StopTracingAsync(bool saveTrace, string scenarioId)
    {
        if (!_tracingStarted)
        {
            return null;
        }

        _tracingStarted = false;

        if (!saveTrace)
        {
            await Context.Tracing.StopAsync();
            return null;
        }

        var tracePath = Path.Combine(Settings.ArtifactsDirectory, $"{scenarioId}-trace.zip");
        await Context.Tracing.StopAsync(new TracingStopOptions
        {
            Path = tracePath
        });
        return tracePath;
    }

    /// <inheritdoc />
    public async ValueTask DisposeAsync()
    {
        await DisposeQuietlyAsync(async () => await Page.CloseAsync());
        await DisposeQuietlyAsync(async () => await Context.CloseAsync());
        await DisposeQuietlyAsync(async () => await Browser.CloseAsync());
        Playwright.Dispose();
    }

    private static async Task DisposeQuietlyAsync(Func<Task> action)
    {
        try
        {
            await action();
        }
        catch
        {
        }
    }
}
